modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/SortNode.java

```diff
@@ -59,18 +58,19 @@ public class SortNode<RowT> extends AbstractNode<RowT> implements SingleNode<Row
      */
     public SortNode(ExecutionContext<RowT> ctx,
             Comparator<RowT> comp,
-            @Nullable Supplier<Integer> offset,
-            @Nullable Supplier<Integer> fetch) {
+            long offset,
+            long fetch) {
         super(ctx);
-        assert fetch == null || fetch.get() >= 0;
-        assert offset == null || offset.get() >= 0;
 
-        limit = fetch == null ? -1 : fetch.get() + (offset == null ? 0 : offset.get());
+        assert fetch == -1 || fetch >= 0;
+        assert offset >= 0;
 
-        if (limit < 1) {
+        limit = fetch == -1 ? -1 : IgniteMath.addExact(fetch, offset);
+
+        if (limit < 1 || limit > Integer.MAX_VALUE) {
```

Comment:
Both queue implementation are backed by arrays.
So, limit greater `Integer.MAX_VALUE` may lead to a int overflow and array out bounds.

Reply:
May lead of course, but we might be lucky